src/Gui/Action.cpp

```diff
@@ -183,21 +192,135 @@ QString Action::statusTip() const
 void Action::setText(const QString & s)
 {
     _action->setText(s);
+    if (_title.isEmpty())
+        setToolTip(_tooltip);
 }
 
 QString Action::text() const
 {
     return _action->text();
 }
 
-void Action::setToolTip(const QString & s)
+void Action::setToolTip(const QString & s, const QString & title)
+{
+    _tooltip = s;
+    _title = title;
+    _action->setToolTip(createToolTip(s,
+                title.isEmpty() ? _action->text() : title, 
+                _action->font(),
+                _action->shortcut().toString(QKeySequence::NativeText),
+                this));
+}
+
+static QString & actionTitle(QString & title)

```

Comment:
In modern C++ it's recommended to put global functions into anonymous namespace, i.e.:
```
namespace {
...
}
```